### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by keliumJU**
**This module contains the sample code snippets for the themes:**

- Introduction to TestingTesting Contexts
- Testing Contexts
- Testing Controllers

_It was explained a little how tests work in phoenix, how ExUnit's ConnCase extends the differences between testing routes with html and json, particularities such as action_fallback. in addition to different commands to test by files, modules, or particular test cases. They even highlighted the difference between testing contexts and schemas (Ecto code)_

### CODE REVIEW

File: lib/hello_web/controllers/fallback_controller.ex [hunk 1]
```diff
@@ -0,0 +1,24 @@
+defmodule HelloWeb.FallbackController do
```
Comment (juanhurtado10) [inline lib/hello_web/controllers/fallback_controller.ex:1]: @keliumJU what is the purpose of a fallback controller?

File: lib/hello_web/controllers/fallback_controller.ex [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    |> put_status(:unprocessable_entity)
+    |> put_view(HelloWeb.ChangesetView)
+    |> render("error.json", changeset: changeset)
+  end
+
+  # This clause is an example of how to handle resources that cannot be found.
+  def call(conn, {:error, :not_found}) do
+    conn
+    |> put_status(:not_found)
+    |> put_view(HelloWeb.ErrorView)
```
Comment (juanhurtado10) [inline lib/hello_web/controllers/fallback_controller.ex:21]: @keliumJU what does `put_view` do?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/hello_web/controllers/fallback_controller.ex (hunk 1 above)
Comment (keliumJU) [reply lib/hello_web/controllers/fallback_controller.ex:1]: Fallback controller what is intended is that developers avoid the repetition of code, in this case we control the error return that should not be explicit, by making use of the action_fallback it is enough to perform the error return.

File: lib/hello_web/controllers/fallback_controller.ex (hunk 2 above)
Comment (keliumJU) [reply lib/hello_web/controllers/fallback_controller.ex:21]: put_view prepares a view to be rendered, we can choose any of our available views with put_view and then with the pipe operator |> pass the necessary parameters

**Review by juanhurtado10 — APPROVED** [review]

**Pull request merged**